src/move/rewards/sources/emojicoin_dot_fun_arena.move

```diff
@@ -0,0 +1,614 @@
+module rewards::emojicoin_dot_fun_arena {
+
+    use aptos_framework::account::{Self, SignerCapability};
+    use aptos_framework::aptos_account;
+    use aptos_framework::aptos_coin::AptosCoin;
+    use aptos_framework::coin::{Self, Coin};
+    use aptos_framework::randomness::Self;
+    use aptos_framework::timestamp;
+    use aptos_std::smart_table::{Self, SmartTable};
+    use emojicoin_dot_fun::emojicoin_dot_fun::{
+        Self,
+        MarketMetadata,
+        market_view,
+        registry_view,
+        unpack_market_metadata,
+        unpack_registry_view
+    };
+    use std::option::Self;
+    use std::signer;
+
+    /// Signer does not correspond to rewards account.
+    const E_NOT_REWARDS: u64 = 0;
+
+    /// Resource account address seed for the registry.
+    const REGISTRY_SEED: vector<u8> = b"Arena registry";
+
+    const U64_MAX: u64 = 0xffffffffffffffff;
+
+    const LOCK_IN_PERIOD_HOURS: u64 = 12;
+    const MELEE_DURATION_HOURS: u64 = 36;
+    const MICROSECONDS_PER_HOUR: u64 = 3_600_000_000;
+
+    /// Flat integrator fee.
+    const INTEGRATOR_FEE_RATE_BPS: u8 = 100;
+    const INTEGRATOR_FEE_RATE_BPS_DUAL_ROUTE: u8 = 50;
+    const REWARDS_PER_MELEE: u64 = 1500 * 100_000_000;
+    const MATCH_PERCENTAGE: u64 = 20;
+    const MAX_PERCENTAGE: u64 = 100;
+    const MAX_MATCH_AMOUNT: u64 = 100_000_000;
+
+    struct Melee has copy, drop, store {
+        /// 1-indexed for conformity with market ID indexing.
+        melee_id: u64,
+        /// Market with lower market ID first.
+        market_metadatas: vector<MarketMetadata>,
+        available_rewards_in_octas: u64,
+        /// In microseconds.
+        start_time: u64,
+        lock_in_period: u64,
+        duration: u64
+    }
+
+    struct Nil has store {}
+
+    struct MeleeEscrow<phantom Emojicoin0, phantom EmojicoinLP0, phantom Emojicoin1, phantom EmojicoinLP1> has key {
+        melee_id: u64,
+        emojicoin_0: Coin<Emojicoin0>,
+        emojicoin_1: Coin<Emojicoin1>,
+        tap_out_fee: u64
+    }
+
+    struct Registry has key {
+        /// Map from melee serial ID to the melee.
+        melees_by_id: SmartTable<u64, Melee>,
+        /// Map from a sorted combination of market IDs (lower ID first) to the melee serial ID.
+        melees_by_market_combo_sorted_market_ids: SmartTable<vector<u64>, u64>,
+        /// Approves transfers from the vault.
+        signer_capability: SignerCapability
+    }
+
+    struct UserMelees has key {
+        /// Set of serial IDs of all melees the user has entered.
+        melee_ids: SmartTable<u64, Nil>
+    }
+
```

Comment:
Useful view functions:

1. User's balance in an escrow
1. User's PnL in an escrow
1. Current melee information
1. Economic parameters
1. Vault state

Reply:
Marking resolved, captured in Linear to-dos